pkg/apis/extensions/v1alpha1/types.go

```diff
@@ -0,0 +1,125 @@
+	DefaultSpec `json:",inline"`
+
+	// Units is a list of unit for the operating system configuration (usually, a systemd unit).
+	// +optional
+	Units []Unit `json:"units,omitempty"`
+	// Files is a list of files that should get written to the host's file system.
+	// +optional
+	Files []File `json:"files,omitempty"`
+}
+
+// Unit is a unit for the operating system configuration (usually, a systemd unit).
+type Unit struct {
+	// Name is the name of a unit.
+	Name string `json:"name"`
+	// Command is the unit's command.
+	// +optional
+	Command *string `json:"command,omitempty"`
+	// Enable describes whether the unit is enabled or not.
+	// +optional
+	Enable *bool `json:"enable,omitempty"`
+	// Content is the unit's content.
+	// +optional
+	Content *string `json:"content,omitempty"`
+	// DropIns is a list of drop-ins for this unit.
+	// +optional
+	DropIns []DropIn `json:"dropIns,omitempty"`
+}
+
+// DropIn is a drop-in configuration for a systemd unit.
+type DropIn struct {
+	// Name is the name of the drop-in.
+	Name string `json:"name"`
+	// Content is the content of the drop-in.
+	Content string `json:"content"`
+}
+
+// File is a file that should get written to the host's file system. The content can either be inlined or
+// referenced from a secret in the same namespace.
+type File struct {
+	// Path is the path of the file system where the file should get written to.
+	Path string `json:"path"`
+	// Permissions describes with which permissions the file should get written to the file system.
+	Permissions string `json:"permissions"`
+	// Content describe the file's content.
+	Content FileContent `json:"content"`
+}
+
+// FileContent can either reference a secret or contain inline configuration.
+type FileContent struct {
+	// SecretRef is a struct that contains information about the referenced secret.
+	// +optional
+	SecretRef *FileContentSecretRef `json:"secretRef,omitempty"`
+	// Inline is a struct that contains information about the inlined data.
+	// +optional
+	Inline *FileContentInline `json:"inline,omitempty"`
+}
+
+// FileContentSecretRef contains keys for referencing a file content's data from a secret in the same namespace.
+type FileContentSecretRef struct {
+	// Name is the name of the secret.
+	Name string `json:"name"`
+	// DataKey is the key in the secret's `.data` field that should be read.
+	DataKey string `json:"dataKey"`
+}
+
+// FileContentInline contains keys for inlining a file content's data and encoding.
+type FileContentInline struct {
+	// Encoding is the file's encoding (e.g. base64).
+	Encoding string `json:"encoding"`
+	// Data is the file's data.
+	Data string `json:"data"`
+}
+
+// OperatingSystemConfigStatus is the status for a OperatingSystemConfig resource
+type OperatingSystemConfigStatus struct {
+	DefaultStatus `json:",inline"`
+
+	// CloudConfig is the generated output for the given operating system configuration in the spec.
+	CloudConfig string `json:"cloudConfig"`
```

Comment:
Gardener needs to read this information and distribute it to others later (e.g., MCM). How can it do this in that case?

Reply:
I would expect that data to be stored in a `Secret`.

Reply:
I'd be fine with either a `Secret` or when this field is moved to the `spec` section. We shouldn't use information that's in a `status` section for further reconciliations.

Reply:
I get the point with using a `Secret` (because it might contain "confidental" information) - although I don't like the indirection.

> We shouldn't use information that's in a status section for further reconciliations.

I totally don't understand that comment. This is an essential part of Kubernetes itself and also of the extensibility concept. This will definitely happen. Many controllers are built this way:
1. Node-Lifecycle controller checks `.status` in `Node` objects written by kubelets
1. ReplicaSet controller acts based on `.status` in `Pod` objects
...
many more

Reply:
>Node-Lifecycle controller checks .status in Node objects written by kubelets
>ReplicaSet controller acts based on .status in Pod objects
>...
>many more
Ok, I guess my wording here was bad: I didn't mean that you cannot use the `status` for the flow of your reconciliation (like your examples do), but they _never_ copy around `status` information to a `spec` section of something else.